src/intersection-observer-polyfill.js

```diff
@@ -0,0 +1,275 @@
+      elementRectNorm = opt_container
+          ? moveLayoutRect(elementRect, opt_container.left, opt_container.top)
+          : elementRect;
+    } else {
+      // If the element is not an AMP element, always assume its position to
+      // the page won't change.
+      if (!this.elementRectForNonAMP_) {
+        const elementRect = LayoutRectFromDomRect(
+            this.element_./*OK*/getBoundingClientRect());
+        this.elementRectForNonAMP_ =
+            moveLayoutRect(elementRect, opt_container.left, opt_container.top);
+      }
+      elementRectNorm = this.elementRectForNonAMP_;
+    }
+
+    // Normalize container element
+    const changeEntry = this.getValidIntersectionChangeEntry_(
+        elementRectNorm, viewport, opt_container);
+    if (changeEntry) {
+      this.callback_(changeEntry);
+    }
+  }
+
+  /**
+   * Return a change entry for that should be compatible with
+   * IntersectionObserverEntry if it's valid with current config.
+   *
+   * Mutates passed in rootBounds to have x and y according to spec.
+   *
+   * @param {!./layout-rect.LayoutRectDef} element element's rect
+   * @param {!./layout-rect.LayoutRectDef} viewport viewport's rect.
+   * @return {?IntersectionObserverEntry} A valid change entry.
+   * @private
+   */
+  getValidIntersectionChangeEntry_(element, viewport, opt_container) {
+    // TODO: need to include owner LayoutRect info
+
+    // Building an IntersectionObserverEntry.
+    let intersectionRect = element;
+    if (opt_container) {
+      intersectionRect = rectIntersection(opt_container, element) ||
+          // No intersection.
+          layoutRectLtwh(0, 0, 0, 0);
+    }
+    intersectionRect = rectIntersection(viewport, intersectionRect) ||
+        // No intersection.
+        layoutRectLtwh(0, 0, 0, 0);
+
+    intersectionRect = moveLayoutRect(intersectionRect, -viewport.left,
+        -viewport.top);
+
+    const ratio = intersectionRatio(intersectionRect, element);
+    const newThresholdSlot = this.getThresholdSlot_(ratio);
+    if (newThresholdSlot == this.prevThresholdSlot_) {
+      return null;
+    }
+    this.prevThresholdSlot_ = newThresholdSlot;
+
+    // The element is relative to (0, 0), while the viewport moves. So, we must
+    // adjust.
+    const boundingClientRect = moveLayoutRect(element, -viewport.left,
+        -viewport.top);
+
+    // Now, move the viewport to (0, 0)
+    const rootBounds = moveLayoutRect(viewport, -viewport.left, -viewport.top);
+
+    return /** @type {!IntersectionObserverEntry} */ ({
+      time: performance.now(),
+      rootBounds: DomRectFromLayoutRect(rootBounds),
+      boundingClientRect: DomRectFromLayoutRect(boundingClientRect),
+      intersectionRect: DomRectFromLayoutRect(intersectionRect),
+      intersectionRatio: ratio,
+    });
+  }
+
+  /**
+   * Returns the slot number that the current ratio fills in.
+   * @param {number} ratio Range from [0, 1]
+   * @return {number} Range from [0, threshold.length]
```

Comment:
The return is the sorted index. IE, if you were to insert something in this array, at what index should it be inserted. (In this case, we treat equivalent values already in the index as being smaller than the inserting value).